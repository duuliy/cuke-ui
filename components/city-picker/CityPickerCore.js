import React, { PureComponent } from "react";
import PropTypes from "prop-types";
import cls from "classnames";
import Spin from "../spin";
import Empty from "../empty";

export default class CityPickerCore extends PureComponent {
  static defaultProps = {
    prefixCls: "cuke-city-picker-core",
    disabledGroups: [],
    loading: false,
    cityList: [],
    notFoundContent: <Empty />
  };
  static propTypes = {
    cityList: PropTypes.arrayOf(
      PropTypes.shape({
        group: PropTypes.string.isRequired,
        resources: PropTypes.arrayOf(PropTypes.object)
      })
    ).isRequired,
    disabledGroups: PropTypes.array,
    onCityGroupChange: PropTypes.func,
    onCityChange: PropTypes.func,
    loading: PropTypes.bool,
    tip: PropTypes.any,
    defaultActiveGroup: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number
    ]),
    activeGroup: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    notFoundContent: PropTypes.any
  };
  state = {
    selectedCityGroup:
      this.props.defaultActiveGroup || this.props.activeGroup || 0,
    selectedCityName: this.props.defaultCityName || this.props.cityName || ""
  };

  onCityGroupChange = (selectedCityGroup, index) => () => {
    this.setState({ selectedCityGroup });

    if (this.props.onCityGroupChange) {
      this.props.onCityGroupChange(selectedCityGroup, index);
    }
  };
  onCityChange = selectedCity => () => {
    this.setState({ selectedCityName: selectedCity.name });
    if (this.props.onCityChange) {
      this.props.onCityChange(selectedCity);
    }
  };

  render() {
    const {
      cityList,
      prefixCls,
      className,
      defaultActiveGroup, //eslint-disable-line
      onCityGroupChange, //eslint-disable-line
      onCityChange, //eslint-disable-line
      disabledGroups,
      defaultCityName, //eslint-disable-line
      loading,
      notFoundContent,
      tip,
      ...attr
    } = this.props;
    const { selectedCityGroup, selectedCityName } = this.state;
    const cityGroups =
      cityList.length >= 1 ? cityList.map(({ group }) => group) : [];

    const currentCityList =
      (
        cityList.find(
          (item, i) =>
            item.group === selectedCityGroup || i === selectedCityGroup
        ) || {}
      ).resources || [];

    return (
      <div className={cls(prefixCls, className)} {...attr}>
        <Spin size="large" spinning={loading} tip={tip}>
          <div className={cls(`${prefixCls}-panel`)}>
            {cityGroups.length >= 1 && (
              <div className={cls(`${prefixCls}-panel-header`)}>
                <ul className={cls(`${prefixCls}-panel-header-wrap`)}>
                  {cityGroups.map((cityGroup, i) => {
                    const isDisabled = disabledGroups.some(
                      group => group === cityGroup || group === i
                    );
                    return (
                      <li
                        onClick={
                          isDisabled
                            ? undefined
                            : this.onCityGroupChange(cityGroup, i)
                        }
                        className={cls(`${prefixCls}-item`, {
                          [`${prefixCls}-active`]:
                            !isDisabled &&
                            (selectedCityGroup === cityGroup ||
                              selectedCityGroup === i),
                          [`${prefixCls}-disabled`]: isDisabled
                        })}
                        key={i}
                      >
                        <span>{cityGroup}</span>
                      </li>
                    );
                  })}
                </ul>
              </div>
            )}
            <div className={cls(`${prefixCls}-panel-content`)}>
              <ul className={cls(`${prefixCls}-panel-content-wrap`)}>
                {cityList.length >= 1
                  ? currentCityList.length >= 1
                    ? currentCityList.map((city, i) => {
                        return (
                          <li
                            className={cls(`${prefixCls}-city`, {
                              [`${prefixCls}-city-selected`]:
                                selectedCityName === city.name
                            })}
                            key={i}
                            onClick={this.onCityChange(city)}
                          >
                            {city.name}
                          </li>
                        );
                      })
                    : notFoundContent
                  : notFoundContent}
              </ul>
            </div>
          </div>
        </Spin>
      </div>
    );
  }
}
