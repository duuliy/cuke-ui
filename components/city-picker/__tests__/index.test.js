import assert from "power-assert";
import React from "react";
import { render, shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";
import CityPicker from "../index";
import Input from "../../input";
import CityPickerCore from "../CityPickerCore";
import Spin from "../../spin";
import Empty from "../../empty";
import Button from "../../button";

const cityList = [
  {
    group: "热门",
    resources: [
      {
        id: 1,
        name: "成都"
      }
    ]
  },
  {
    group: "ABCDE",
    resources: [
      {
        id: 1,
        name: "成都"
      },
      {
        id: 2,
        name: "资阳"
      }
    ]
  },
  {
    group: "FGHIJ",
    resources: [
      {
        id: 3,
        name: "上海"
      },
      {
        id: 4,
        name: "乐至"
      }
    ]
  },
  {
    group: "KLMNO",
    resources: [
      {
        id: 5,
        name: "自贡"
      },
      {
        id: 6,
        name: "北京"
      },
      {
        id: 7,
        name: "香港"
      },
      {
        id: 8,
        name: "南京"
      },
      {
        id: 9,
        name: "简阳"
      }
    ]
  },
  {
    group: "PQRST",
    resources: [
      {
        id: 10,
        name: "广州"
      }
    ]
  },
  {
    group: "UVWXYZ",
    resources: [
      {
        id: 11,
        name: "西安"
      }
    ]
  }
];

describe("<CityPicker/>", () => {
  it("should render CityPicker", () => {
    const wrapper = render(<CityPicker cityList={cityList} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it("should render custom size", () => {
    const wrapper = render(
      <div>
        <CityPicker cityList={cityList} placeholder="small" size="small" />
        <CityPicker cityList={cityList} placeholder="default" size="default" />
        <CityPicker cityList={cityList} placeholder="large" size="large" />
      </div>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should render Input when render CityPicker", () => {
    const wrapper = shallow(<CityPicker cityList={cityList} />);
    assert(wrapper.find(Input).length === 1);
  });

  it("should render CityPickerCore when render CityPicker", () => {
    const wrapper = shallow(<CityPicker cityList={cityList} />);
    assert(wrapper.find(CityPickerCore).length === 1);
  });

  it("should render cuke-city-picker classnames", () => {
    const wrapper = shallow(<CityPicker cityList={cityList} />);
    assert(wrapper.find(".cuke-city-picker").length === 1);
  });

  it("should render cuke-city-picker-core classnames", () => {
    const wrapper = shallow(<CityPickerCore cityList={cityList} />);
    assert(wrapper.find(".cuke-city-picker-core").length === 1);
  });
  it("should render CityPickerCore ", () => {
    const wrapper = shallow(<CityPicker cityList={cityList} />);
    assert(wrapper.find(CityPickerCore).length === 1);
  });

  it("should render custom default active city group", () => {
    const wrapper = shallow(
      <CityPicker cityList={cityList} defaultActiveGroup={2} />
    );
    assert(wrapper.state().selectedCityGroup === 2);
  });

  it("should render custom active city group", () => {
    const wrapper = shallow(<CityPicker cityList={cityList} activeGroup={2} />);
    assert(wrapper.state().selectedCityGroup === 2);
  });

  it("should render custom default city name", () => {
    const wrapper = shallow(
      <CityPicker cityList={cityList} defaultCityName={"成都"} />
    );
    assert(wrapper.state().selectedCityName === "成都");
  });

  it("should render custom city name", () => {
    const wrapper = shallow(
      <CityPicker cityList={cityList} cityName={"成都"} />
    );
    assert(wrapper.state().selectedCityName === "成都");
  });

  it("should render custom string active city group ", () => {
    const wrapper = shallow(
      <CityPicker cityList={cityList} defaultActiveGroup={"热门"} />
    );
    assert(wrapper.state().selectedCityGroup === "热门");
  });

  it("should render custom disabled city group and can not trigger click event ", () => {
    const onChange = jest.fn();
    const wrapper = shallow(
      <CityPicker
        onCityGroupChange={onChange}
        cityList={cityList}
        disabledGroups={[0, 1, 2, 3, 4, 5, 6]}
      />
    );
    wrapper.find(".cuke-city-picker").simulate("click");
    expect(onChange).not.toHaveBeenCalled();
  });

  it("should can trigger city group change event ", () => {
    const onChange = jest.fn();
    const wrapper = shallow(
      <CityPicker onCityGroupChange={onChange} cityList={cityList} />
    );
    setTimeout(() => {
      wrapper.find(".cuke-city-picker-core-item").simulate("click");
      expect(onChange).toHaveBeenCalled();
    });
  });

  it("should can trigger city group change event ", () => {
    const onChange = jest.fn();
    const wrapper = shallow(
      <CityPicker onCityChange={onChange} cityList={cityList} />
    );
    setTimeout(() => {
      wrapper.find(".cuke-city-picker-core-city").simulate("click");
      expect(onChange).toHaveBeenCalled();
    });
  });

  it("should can trigger city group change event in CityPickerCore ", () => {
    const onChange = jest.fn();
    const wrapper = shallow(
      <CityPickerCore onCityChange={onChange} cityList={cityList} />
    );
    wrapper.find(".cuke-city-picker-core-city").simulate("click");
    expect(onChange).toHaveBeenCalled();
  });

  it("should can trigger city group change event in CityPickerCore ", () => {
    const onChange = jest.fn();
    const wrapper = shallow(
      <CityPickerCore onCityGroupChange={onChange} cityList={cityList} />
    );
    wrapper
      .find(".cuke-city-picker-core-item")
      .at(0)
      .simulate("click");
    expect(onChange).toHaveBeenCalled();
  });

  it("should window cannot trigger click event when un mount", () => {
    const onClick = jest.fn();
    const wrapper = shallow(<CityPicker cityList={cityList} />);
    wrapper.unmount();
    window.onclick = () => onClick;
    expect(onClick).not.toHaveBeenCalled();
  });

  it("should trigger panel visible change", () => {
    const onPanelVisibleChange = jest.fn();
    const wrapper = shallow(
      <CityPicker
        cityList={cityList}
        onPanelVisibleChange={onPanelVisibleChange}
      />
    );
    wrapper.find(".cuke-city-picker-input").simulate("click");
    expect(onPanelVisibleChange).toHaveBeenCalled();
  });

  it("should trigger onCityChange by CityPickerCore", () => {
    const onCityChange = jest.fn();
    const wrapper = shallow(
      <CityPickerCore cityList={cityList} onCityChange={onCityChange} />
    );
    wrapper
      .find(".cuke-city-picker-core-city")
      .at(0)
      .simulate("click");
    expect(onCityChange).toHaveBeenCalled();
  });

  it("should render spin when loading is true", () => {
    const wrapper = shallow(
      <CityPickerCore cityList={cityList} loading={true} />
    );
    assert(wrapper.find(Spin).length === 1);
  });

  it("should render spin when loading is true", () => {
    const wrapper = shallow(<CityPicker cityList={cityList} loading={true} />);
    assert(wrapper.find(Spin).length === 0);
  });

  it("should trigger onCityGroupChange by CityPickerCore", () => {
    const onCityGroupChange = jest.fn();
    const wrapper = shallow(
      <CityPickerCore
        cityList={cityList}
        onCityGroupChange={onCityGroupChange}
      />
    );
    wrapper
      .find(".cuke-city-picker-core-item")
      .at(0)
      .simulate("click");
    expect(onCityGroupChange).toHaveBeenCalled();
  });

  it("should cannot trigger panel visible change when disabled groups", () => {
    const onPanelVisibleChange = jest.fn();
    const wrapper = shallow(
      <CityPicker
        cityList={cityList}
        disabled
        onPanelVisibleChange={onPanelVisibleChange}
      />
    );
    wrapper.find(".cuke-city-picker").simulate("click");
    expect(onPanelVisibleChange).not.toHaveBeenCalled();
  });

  it("should render not found content when cityList is Empty", () => {
    const wrapper = shallow(<CityPickerCore cityList={[]} />);
    assert(wrapper.find(Empty).length === 1);
    assert(wrapper.find(".cuke-city-picker-panel-header").length === 0);
  });

  it("should render custom not found content", () => {
    const wrapper = shallow(
      <CityPickerCore cityList={[]} notFoundContent={<Button>111</Button>} />
    );
    assert(wrapper.find(Button).length === 1);
  });

  it.skip("should emit onCityGroupChange when group clicked ", () => {
    const onChange = jest.fn();
    const wrapper = mount(
      <CityPicker onCityGroupChange={onChange} cityList={cityList} />
    );
    wrapper.setState({
      visible: true
    });
    wrapper
      .find(".cuke-city-picker-core-item")
      .at(1)
      .simulate("click");
    expect(onChange).toHaveBeenCalled();
  });

  it.skip("should emit onCityChange when group clicked ", () => {
    const onChange = jest.fn();
    const wrapper = mount(
      <CityPicker onCityChange={onChange} cityList={cityList} />
    );
    wrapper.setState({
      visible: true
    });
    wrapper
      .find(".cuke-city-picker-core-city")
      .at(0)
      .simulate("click");
    expect(onChange).toHaveBeenCalled();
  });

  it.skip("should emit onClear ", () => {
    const onChange = jest.fn();
    const wrapper = mount(
      <CityPicker
        allowClear
        onCityChange={onChange}
        cityList={cityList}
        selectedCityName="成都"
      />
    );
    wrapper.setState({ selectedCityName: "成都" });
    wrapper
      .find(".cuke-input-clear")
      .at(0)
      .simulate("click");
    expect(wrapper.state().selectedCityName).toBe("");
    expect(onChange).toHaveBeenCalled();
  });

  // it("should cannot trigger panel visible change when disabled groups", () => {
  //   const onPanelVisibleChange = jest.fn();
  //   const wrapper = mount(
  //     <CityPicker
  //       cityList={cityList}
  //       onPanelVisibleChange={onPanelVisibleChange}
  //     />
  //   );
  //   wrapper.setState({visible: true})
  //   wrapper.find(".cuke-city-picker-core-item").at(0).simulate("click");
  //   expect(onPanelVisibleChange).toHaveBeenCalled();
  // });

  it("should find custom popup container class name ", () => {
    const wrapper = shallow(<CityPicker popupContainerClassName="test" />);
    assert(wrapper.find(".test").length === 1);
  });
});
