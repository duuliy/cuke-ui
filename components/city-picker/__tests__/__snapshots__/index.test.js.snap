// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CityPicker/> should render CityPicker 1`] = `
<div
  class="cuke-city-picker"
>
  <div
    class="cuke-city-picker-inner"
  >
    <div
      class="cuke-input-wrapper cuke-input-has-suffix"
    >
      <input
        class="cuke-input cuke-city-picker-input"
        placeholder="请选择"
        readonly=""
        type="text"
        value=""
      />
      <span
        class="cuke-input-suffix"
      >
        <svg
          class="cuke-city-picker-arrow"
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
        >
          <polyline
            points="6 9 12 15 18 9"
          />
        </svg>
      </span>
    </div>
  </div>
  <div
    class="cuke-city-picker-content cuke-city-picker-close cuke-ui-no-animate"
  >
    <div
      class="cuke-city-picker-core"
    >
      <div
        class="cuke-city-picker-core-panel"
      >
        <div
          class="cuke-city-picker-core-panel-header"
        >
          <ul
            class="cuke-city-picker-core-panel-header-wrap"
          >
            <li
              class="cuke-city-picker-core-item cuke-city-picker-core-active"
            >
              <span>
                热门
              </span>
            </li>
            <li
              class="cuke-city-picker-core-item"
            >
              <span>
                ABCDE
              </span>
            </li>
            <li
              class="cuke-city-picker-core-item"
            >
              <span>
                FGHIJ
              </span>
            </li>
            <li
              class="cuke-city-picker-core-item"
            >
              <span>
                KLMNO
              </span>
            </li>
            <li
              class="cuke-city-picker-core-item"
            >
              <span>
                PQRST
              </span>
            </li>
            <li
              class="cuke-city-picker-core-item"
            >
              <span>
                UVWXYZ
              </span>
            </li>
          </ul>
        </div>
        <div
          class="cuke-city-picker-core-panel-content"
        >
          <ul
            class="cuke-city-picker-core-panel-content-wrap"
          >
            <li
              class="cuke-city-picker-core-city"
            >
              成都
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<CityPicker/> should render custom size 1`] = `
<div>
  <div
    class="cuke-city-picker"
  >
    <div
      class="cuke-city-picker-inner"
    >
      <div
        class="cuke-input-wrapper cuke-input-has-suffix"
      >
        <input
          class="cuke-input cuke-city-picker-input cuke-input-small"
          placeholder="small"
          readonly=""
          type="text"
          value=""
        />
        <span
          class="cuke-input-suffix"
        >
          <svg
            class="cuke-city-picker-arrow"
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1em"
          >
            <polyline
              points="6 9 12 15 18 9"
            />
          </svg>
        </span>
      </div>
    </div>
    <div
      class="cuke-city-picker-content cuke-city-picker-close cuke-ui-no-animate"
    >
      <div
        class="cuke-city-picker-core"
      >
        <div
          class="cuke-city-picker-core-panel"
        >
          <div
            class="cuke-city-picker-core-panel-header"
          >
            <ul
              class="cuke-city-picker-core-panel-header-wrap"
            >
              <li
                class="cuke-city-picker-core-item cuke-city-picker-core-active"
              >
                <span>
                  热门
                </span>
              </li>
              <li
                class="cuke-city-picker-core-item"
              >
                <span>
                  ABCDE
                </span>
              </li>
              <li
                class="cuke-city-picker-core-item"
              >
                <span>
                  FGHIJ
                </span>
              </li>
              <li
                class="cuke-city-picker-core-item"
              >
                <span>
                  KLMNO
                </span>
              </li>
              <li
                class="cuke-city-picker-core-item"
              >
                <span>
                  PQRST
                </span>
              </li>
              <li
                class="cuke-city-picker-core-item"
              >
                <span>
                  UVWXYZ
                </span>
              </li>
            </ul>
          </div>
          <div
            class="cuke-city-picker-core-panel-content"
          >
            <ul
              class="cuke-city-picker-core-panel-content-wrap"
            >
              <li
                class="cuke-city-picker-core-city"
              >
                成都
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="cuke-city-picker"
  >
    <div
      class="cuke-city-picker-inner"
    >
      <div
        class="cuke-input-wrapper cuke-input-has-suffix"
      >
        <input
          class="cuke-input cuke-city-picker-input"
          placeholder="default"
          readonly=""
          type="text"
          value=""
        />
        <span
          class="cuke-input-suffix"
        >
          <svg
            class="cuke-city-picker-arrow"
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1em"
          >
            <polyline
              points="6 9 12 15 18 9"
            />
          </svg>
        </span>
      </div>
    </div>
    <div
      class="cuke-city-picker-content cuke-city-picker-close cuke-ui-no-animate"
    >
      <div
        class="cuke-city-picker-core"
      >
        <div
          class="cuke-city-picker-core-panel"
        >
          <div
            class="cuke-city-picker-core-panel-header"
          >
            <ul
              class="cuke-city-picker-core-panel-header-wrap"
            >
              <li
                class="cuke-city-picker-core-item cuke-city-picker-core-active"
              >
                <span>
                  热门
                </span>
              </li>
              <li
                class="cuke-city-picker-core-item"
              >
                <span>
                  ABCDE
                </span>
              </li>
              <li
                class="cuke-city-picker-core-item"
              >
                <span>
                  FGHIJ
                </span>
              </li>
              <li
                class="cuke-city-picker-core-item"
              >
                <span>
                  KLMNO
                </span>
              </li>
              <li
                class="cuke-city-picker-core-item"
              >
                <span>
                  PQRST
                </span>
              </li>
              <li
                class="cuke-city-picker-core-item"
              >
                <span>
                  UVWXYZ
                </span>
              </li>
            </ul>
          </div>
          <div
            class="cuke-city-picker-core-panel-content"
          >
            <ul
              class="cuke-city-picker-core-panel-content-wrap"
            >
              <li
                class="cuke-city-picker-core-city"
              >
                成都
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="cuke-city-picker"
  >
    <div
      class="cuke-city-picker-inner"
    >
      <div
        class="cuke-input-wrapper cuke-input-has-suffix"
      >
        <input
          class="cuke-input cuke-city-picker-input cuke-input-large"
          placeholder="large"
          readonly=""
          type="text"
          value=""
        />
        <span
          class="cuke-input-suffix"
        >
          <svg
            class="cuke-city-picker-arrow"
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1em"
          >
            <polyline
              points="6 9 12 15 18 9"
            />
          </svg>
        </span>
      </div>
    </div>
    <div
      class="cuke-city-picker-content cuke-city-picker-close cuke-ui-no-animate"
    >
      <div
        class="cuke-city-picker-core"
      >
        <div
          class="cuke-city-picker-core-panel"
        >
          <div
            class="cuke-city-picker-core-panel-header"
          >
            <ul
              class="cuke-city-picker-core-panel-header-wrap"
            >
              <li
                class="cuke-city-picker-core-item cuke-city-picker-core-active"
              >
                <span>
                  热门
                </span>
              </li>
              <li
                class="cuke-city-picker-core-item"
              >
                <span>
                  ABCDE
                </span>
              </li>
              <li
                class="cuke-city-picker-core-item"
              >
                <span>
                  FGHIJ
                </span>
              </li>
              <li
                class="cuke-city-picker-core-item"
              >
                <span>
                  KLMNO
                </span>
              </li>
              <li
                class="cuke-city-picker-core-item"
              >
                <span>
                  PQRST
                </span>
              </li>
              <li
                class="cuke-city-picker-core-item"
              >
                <span>
                  UVWXYZ
                </span>
              </li>
            </ul>
          </div>
          <div
            class="cuke-city-picker-core-panel-content"
          >
            <ul
              class="cuke-city-picker-core-panel-content-wrap"
            >
              <li
                class="cuke-city-picker-core-city"
              >
                成都
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
